cloud-controller-manager/do/kubernetes_lb_validator.go

```diff
@@ -0,0 +1,165 @@
+	"fmt"
+	"net/http"
+
+	"github.com/digitalocean/godo"
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// LBService represents the v1.service lb object data
+type LBService struct {
+	APIVersion string
+	Kind       string
+	Metadata   Metadata
+	Spec       map[string]interface{}
+	Selector   map[string]interface{}
+	Type       string
+}
+
+// Metadata represents the metadata field in the lb service object
+type Metadata struct {
+	Annotations interface{} `json:"annotation data,omitempty"`
+	Name        string
+	Namespace   string
+}
+
+// DOKSLBServiceValidator validates service type LB
+type KubernetesLBServiceValidator struct {
+	decoder    *admission.Decoder
+	Log        logr.Logger
+	GClient    *godo.Client
+	DropletTag string
+	Region     string
+}
+
+// DOKSLBServiceValidator ...
+func (v *KubernetesLBServiceValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	svc := &v1.Service{}
+
+	v.Log.V(6).Info("decoding received request")
+	err := v.decoder.Decode(req, svc)
+	if err != nil {
+		v.Log.Error(err, "failed to decode request")
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("failed to decode request: %v", err))
+	}
+
+	v.Log.V(6).Info("checking received request")
+	if svc.Spec.Type != v1.ServiceTypeLoadBalancer || svc.DeletionTimestamp != nil {
+		return admission.Allowed("ignoring the service which is either not a load balancer or is being deleted")
+	}
+
+	// TODO: these forwarding rules are a placeholder. Further development is required to extract the values from the svc object
+	forwardingRules := []godo.ForwardingRule{
+		{
+			EntryProtocol:  "http",
+			EntryPort:      80,
+			TargetProtocol: "http",
+			TargetPort:     80,
+			CertificateID:  "",
+			TlsPassthrough: false,
+		}}
+
+	lbRequest := v.buildRequest(svc.Name, v.Region, v.DropletTag, forwardingRules)
+
+	var resp *godo.Response
+
+	if req.OldObject.Object == nil {
+		if len(svc.Status.LoadBalancer.Ingress) == 0 || svc.Status.LoadBalancer.Ingress[0].IP == "" {
+			resp, err = v.validateCreate(ctx, lbRequest)
+			if err != nil {
+				if v.checkError(resp) {
+					v.Log.Error(err, "failed to validate lb creation")
+					return admission.Denied(fmt.Sprintf("failed to validate lb creation: %v", err))
+				}
+				v.Log.Error(err, "failed to validate lb creation")
+				return admission.Errored(int32(resp.StatusCode), err)
+			}
+			return admission.Allowed("valid lb create request")
+		}
+	}
+
+	err = v.decoder.DecodeRaw(req.OldObject, svc)
+	if err != nil {
+		v.Log.Error(err, "failed to decode existing object")
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("failed to decode request: %v", err))
+	}
+
+	currentLBID := svc.Annotations["kubernetes.digitalocean.com/load-balancer-id"]
+	if currentLBID == "" {
+		v.Log.Error(fmt.Errorf("missing loadbalancer id in annotation"), "missing kubernetes.digitalocean.com/load-balancer-id annotation")
```

Comment:
This should not be considered an error: the old service object might not be of type `LoadBalancer`; or it might be but the configuration is invalid, or something prevents the annotation from being set (e.g., a conflicting operator/controller in the cluster).